### PR TITLE
Very WIP: Stage2 revival

### PR DESCRIPTION
**Pull request by Keno**
Needs some corresponding changes in Cthulhu

### CODE REVIEW

**Comment by Keno** [comment]
> Needs some corresponding changes in Cthulhu

https://github.com/JuliaDebug/Cthulhu.jl/pull/303

**Comment by Keno** [comment]
Alright, I think at this point, I've proven out that all the compiler interfaces do now support what we need. I think the big next thing here is to unify the codegen between stage1 and stage2 and then rewrite the stage2 inference support to be less awful, but I think there's a reasonably clear path here.

File: src/stage2/interpreter.jl [hunk 1]
```diff
@@ -35,22 +40,161 @@ struct ADInterpreter <: AbstractInterpreter
...
 raise_level(interp::ADInterpreter) = change_level(interp, interp.current_level + 1)
 lower_level(interp::ADInterpreter) = change_level(interp, interp.current_level - 1)
 
+Cthulhu.get_optimized_code(interp::ADInterpreter, curs::ADCursor) = interp.opt[curs.level][curs.mi]
+Cthulhu.AbstractCursor(interp::ADInterpreter, mi::MethodInstance) = ADCursor(0, mi)
+
+# This is a lie, but let's clean this up later
+Cthulhu.can_descend(interp::ADInterpreter, mi::MethodInstance, optimize::Bool) = true
+
+function Cthulhu.lookup(interp::ADInterpreter, curs::ADCursor, optimize::Bool; allow_no_src::Bool=false)
```
Comment (aviatesk) [inline src/stage2/interpreter.jl:60]:
```suggestion
function Cthulhu.lookup(interp::ADInterpreter, curs::ADCursor, optimize::Bool)
```
No need to overload the kwarg method (it's only used in the Cthulhu test suite).

Comment (aviatesk) [inline src/stage2/interpreter.jl:60]: We also need to overload `Cthulhu.lookup_constproped(interp::AbstractInterpreter, curs::AbstractCursor, override::InferenceResult, optimize::Bool)` to support constprop'ed callsite.

**Comment by Keno** [comment]
My plan is to merge this without hooking this up and then to continue working on this on main.

**Pull request merged**